nixos/modules/services/search/opensearch.nix

```diff
@@ -0,0 +1,221 @@
+{ config, lib, pkgs, ... }:
+
+with lib;
+
+let
+  cfg = config.services.opensearch;
+
+  opensearchConfig = ''
+    network.host: ${cfg.listenAddress}
+    cluster.name: ${cfg.cluster_name}
+    ${lib.optionalString cfg.single_node "discovery.type: single-node"}
+
+    http.port: ${toString cfg.port}
+    transport.port: ${toString cfg.tcp_port}
+
+    ${cfg.extraConf}
+  '';
+
+  configDir = cfg.dataDir + "/config";
+
+  opensearchYml = pkgs.writeTextFile {
+    name = "opensearch.yml";
+    text = opensearchConfig;
+  };
+
+  loggingConfigFilename = "log4j2.properties";
+  loggingConfigFile = pkgs.writeTextFile {
+    name = loggingConfigFilename;
+    text = cfg.logging;
+  };
+in
+{
+
+  ###### interface
+
+  options.services.opensearch = {
+    enable = mkOption {
+      description = lib.mdDoc "Whether to enable OpenSearch.";
+      default = false;
+      type = types.bool;
+    };
+
+    package = mkOption {
+      description = lib.mdDoc "OpenSearch package to use.";
+      default = pkgs.opensearch;
+      defaultText = literalExpression "pkgs.opensearch";
+      type = types.package;
+    };
+
+    listenAddress = mkOption {
+      description = lib.mdDoc "OpenSearch listen address.";
+      default = "127.0.0.1";
+      type = types.str;
+    };
+
+    port = mkOption {
+      description = lib.mdDoc "OpenSearch port to listen for HTTP traffic.";
+      default = 9200;
+      type = types.port;
+    };
+
+    tcp_port = mkOption {
+      description = lib.mdDoc "OpenSearch port for the node to node communication.";
+      default = 9300;
+      type = types.int;
+    };
+
+    cluster_name = mkOption {
+      description = lib.mdDoc "OpenSearch name that identifies your cluster for auto-discovery.";
+      default = "opensearch";
+      type = types.str;
+    };
```

Comment:
Put those below `services.opensearch.settings`, so you can later use `format.generate "opensearch.yml cfg.settings`. The options should then use the proper path that the config expects.

Reply:
So we leave cluster name and so. Just set inside enable them into `settings` and use `settings` to generate that file

Reply:
Whatever goes into the config files should be below `settings`.

Reply:
Alright. I will look after Fosdem :)

Reply:
Hey you're going to Fosdem ! I'll be there too :) I'll be sticking the Nix room on Sunday afternoon hit me up !

Reply:
@mweinelt done :)